substrate/frame/contracts/src/tests.rs

```diff
@@ -4249,3 +4249,113 @@ fn gas_consumed_is_linear_for_nested_calls() {
 		assert_eq!(gas_max, gas_0 + gas_per_recursion * max_call_depth as u64);
 	});
 }
+
+#[test]
+fn read_only_call_cannot_store() {
+	let (wasm_caller, _code_hash_caller) = compile_module::<Test>("read_only_call").unwrap();
+	let (wasm_callee, _code_hash_callee) = compile_module::<Test>("store_call").unwrap();
+	ExtBuilder::default().existential_deposit(200).build().execute_with(|| {
+		let _ = <Test as Config>::Currency::set_balance(&ALICE, 1_000_000);
+
+		// Create both contracts: Constructors do nothing.
+		let addr_caller =
+			builder::bare_instantiate(Code::Upload(wasm_caller)).build_and_unwrap_account_id();
+		let addr_callee =
+			builder::bare_instantiate(Code::Upload(wasm_callee)).build_and_unwrap_account_id();
+
+		// Read-only call fails when modifying storage.
+		assert_err_ignore_postinfo!(
+			builder::call(addr_caller).data((&addr_callee, 100u32).encode()).build(),
+			<Error<Test>>::ContractTrapped
```

Comment:
Why in this case is it `ContractTrapped` instead of `StateChangeDenied`?

Reply:
Yes it should be `StateChangeDenied` when looking at the code.

Reply:
We have a flow call_contract_A->A.call(contract B, read_only)->B.set_storage. We encountered a trap in B. Should the error be propagated?

Reply:
Yes you are right. The calling contract does not really learn the reason why a callee had an error. The error gets thrown away. This is the case almost any error unfortunately. So yes this is correct here.